### PR TITLE
Allow specifying multiple namespaces in a single run, and accept pods in "Completed" status as a Good state instead of failure

### PR DESCRIPTION
**Pull request by 9strands**
Added capability to add comma-separated namespaces to the `--namespace` option for a multi-namespace check (necessary for projects where a single application is spread across multiple namespaces for whatever developer reasons are provided) as such:

`--namespace "namespace1,namespace2,namespace3"`

Additional fix added by yvarbanov for acknowledging pods in "Completed" as successful in pods.py - the script was reporting those as failed even though they had successfully finished their run and were terminated and causing the monitor to show failures to show in namespaces where all pods were in good states.

To utilize this function in Icinga2, for example, the following code can be used to apply the monitoring capability to a cluster:

 **Host Entry**
```
object Host "OpenShift K8s Cluster A" {
    check_command = "dummy"
    vars.address = "openshift.example.com"
    vars.bastion_address = "jumphost." + vars.address
    vars.api_address = "api." + vars.address
    vars.dummy_state = 0
    vars.bastion_user = "bastion-user"
    vars.http_port = "6443"
    enable_passive_checks = false
    vars.ocp_infra_namespace_groups = {
         Application1 = "app1web,app1db,app1int"
         Application2 = "app2web,app2db,app2int"
         OCP_Core = "openshift-authentication,openshift-apiserver,openshift-cloud-credential-operator,openshift-machine-api,openshift-config-operator,openshift-console,openshift-cluster-storage-operator"
    }
    vars.client_endpoint = name
    vars.hosttype = "infra_ocp"
    vars.hostenv = "production"
    vars.notification.mail.groups = [ "icingaadmins" ]
}
```

**Service Config**
```
apply Service "OCP Infra Service " for (key => value in host.vars.ocp_infra_namespace_groups) to Host  {
    import "generic_service_babylon_ocp"
    check_command = "check_k8s_pods"
    vars.api_url = "$host.vars.api_address$"
    vars.api_port = "$host.vars.http_port$"
    vars.ocp_namespace = value
    vars.serviceenv = "$host.vars.hostenv$"
    assign where ( match("infra_ocp*", host.vars.hosttype) )
}
```

### CODE REVIEW

**Comment by jacobbaungard** [comment]
Hi @9strands,

Thanks a lot for sending this in. Sounds like sensible changes to me.

I see there's a couple of new lgtm alerts, but they seem to be a lines which just has changed indentation, so we could ignore that for the time being.

However I think we should:

1) Separate out the two different fixes into separate commits
2) Add test cases for both fixes (see example here: https://github.com/ITRS-Group/check_k8s/blob/44647cdc5dd9b4664446bf7e5b1c89e6bef7f5a7/tests/test_pods.py#L53 and here: https://github.com/ITRS-Group/check_k8s/blob/44647cdc5dd9b4664446bf7e5b1c89e6bef7f5a7/tests/test_cli.py#L30)

**Comment by 9strands** [comment]
Hi @jacobbaungard,

  I'm sorry, I'm new to this - I'm unsure how to do what you're asking me to do - how do you split commits out from a PR?

  Also, I have zero idea how to add tests in.  I have reviewed the code which you mention for the tests.  I'm very new to python and I do not know how to setup these tests using the example you have provided.

  Can you point me to any documentation to help?
  

> Hi @9strands,
> 
> Thanks a lot for sending this in. Sounds like sensible changes to me.
> 
> I see there's a couple of new lgtm alerts, but they seem to be a lines which just has changed indentation, so we could ignore that for the time being.
> 
> However I think we should:
> 
>     1. Separate out the two different fixes into separate commits
> 
>     2. Add test cases for both fixes (see example here: https://github.com/ITRS-Group/check_k8s/blob/44647cdc5dd9b4664446bf7e5b1c89e6bef7f5a7/tests/test_pods.py#L53
>         and here: https://github.com/ITRS-Group/check_k8s/blob/44647cdc5dd9b4664446bf7e5b1c89e6bef7f5a7/tests/test_cli.py#L30
>        )

**Comment by 9strands** [comment]
Hi, @jacobbaungard, I found a developer who was willing to sit with me to try to help me understand how to comply with your requests above.  

Accordingly, I've removed the commit of the file which adds the pass-condition for a "Successful" check from pod/resource.py and we attempted to add a test in in a second commit, which is also hopefully what you're looking for.  If this is indeed correct, then, then I'll send a second PR in with the status "Successful" check from pod/resource.py after this one has been approved and merged.

**Comment by jacobbaungard** [comment]
Hi @9strands, that looks better thanks! Sorry I did not get around to answering your previous query with more guidance.

I tested this out a bit, and did find a couple of issues:
- `namespace` is today an optional argument, however the plugin now crashes if no `namespace` argument is provided
- In case of errors, i.e providing a wrong URL, we're not correctly handling the error, (see example below).

current behaviour:
```
# check_k8s.py --resource pods --host invalid_url --token_file ... --namespace mynamespace
UNKNOWN - [Errno -2] Name or service not known
```

After this change:
```
# check_k8s.py --resource pods --host invalid_url --token_file ... --namespace mynamespace
  File "/opt/plugins/check_k8s.py", line 66, in <module>
    main()
  File "/opt/plugins/check_k8s.py", line 57, in main
    output = health_check(response).output
  File "/opt/plugins/k8s/result.py", line 73, in output
    return self._get_output(NaemonState.OK, RESULT_SUCCESS, channel=sys.stdout)
  File "/opt/plugins/k8s/result.py", line 58, in _get_output
    conditions="\n".join(self._messages[state]),
KeyError: <NaemonState.OK: 0>
```
I hope to be able to have time later this week to take a look at this, so we can get this patch in. However if you find a solution to the above in the meantime, that would also be great.

Thanks again for the contribution!

**Comment by 9strands** [comment]
Hi, @jacobbaungard, in thinking about it, I think the loop that I created to allow multiple namespaces likely needs to be put into a conditional if-clause to handle the case where there is no namespace - that is most likely why no namespace being specified fails - no namespace skips the while-loop altogether, meaning none of the setup steps actually occur.  I can think about how to do this in pseudo-code clearly, but I'm still not sure about how to translate it into this code, yet, as I'm not sure about how some of this code hangs-together, yet.  I'll learn more the more I work with it, I'm sure, and I'm also sure there is a more elegant way to accomplish this which I haven't figured out yet.

**Comment by 9strands** [comment]
[jacobbaungard](https://github.com/jacobbaungard) - skipping the namespace argument now returns all namespaces on the cluster, as expected before, I believe.

**Comment by erikdsjostrom** [comment]
Hi @9strands 

This certainly looks like it would solve the issue when namespace is None, but we have ended up with a lot of duplicated code. I think we can remove most of it we split the for loop into two separate loops, one for generating the urls, and for doing the requests. 

You seem interested in learning a bit about this so I'll give you some example code to get you started, let me know if you want some more pointers or if I misjudged your interest completely and you just want me write out exactly what I mean in code. :)

```
    urls = []
    if parsed.namespace is not None:
      for i in parsed.namespace.split(","):
        # Build URL using input arguments
        urls.append(build_url(
            <...>,
            namespace=i
        ))
    else:
        urls.append(build_url(
            <...>
            namespace=None
        ))
```

**Comment by 9strands** [comment]
Hi, @erikdsjostrom - no, you have estimated my desire to learn correctly.  I will see if I can figure out how to implement this.  While I was writing it, the duplication of code bothered me - I did not like it in the least - but I couldn't (at my knowledge-level of this code and of python itself) see how to compress this without rewriting the whole structure of the program.  I know how I would do this as a shell script or as perl, but not python.

I had considered spooling all of the potential url calls into an array, and then looping through them one-at-a-time, but I wasn't sure how that would make the code much cleaner. I'll see if I can piece it together with your hint.

**Comment by 9strands** [comment]
@erikdsjostrom - I believe this commit is in line with what you mentioned?  Sorry for the delay in response, I'm in APAC and I didn't see this until I came back to work on Monday.

File: check_k8s.py [hunk 1]
```diff
@@ -23,45 +23,30 @@ def main():
...
+      urls.append(build_url(
+          host=parsed.host,
+          port=parsed.port,
+          resource=parsed.resource,
+          is_core=is_core,
+          namespace=None
+      ))
       # Request and check health data
+    for url in urls:
+        print(f"{url}\n")
```
Comment (erikdsjostrom) [inline check_k8s.py:49]: Left over print statement?

**Comment by erikdsjostrom** [comment]
That looks exactly like what I was thinking, and no worries about the delay, I'm in no rush.

File: check_k8s.py (hunk 1 above)
Comment (9strands) [reply check_k8s.py:49]: Ah, yes, it is.  I'll remove that.

**Review by erikdsjostrom — APPROVED** [review]
This looks good to me, I'll go ahead and clean up the commits and get this merged unless you have any objections.

**Comment by 9strands** [comment]
No objections from me

**Comment by erikdsjostrom** [comment]
Merged as of [f0dd30f](https://github.com/ITRS-Group/check_k8s/commit/f0dd30f6405fd64ff5a437824dcd5af59753550d)! Thank you very much both reporting and fixing this issue @9strands

**Comment by 9strands** [comment]
Awesome - thanks @erikdsjostrom

**Pull request closed without merging**